File: experimental-plugins/agentic-workflows/src/agentic_workflows/templates/renderer.py
"""Jinja2 template rendering for workflows."""

from __future__ import annotations

from pathlib import Path
from typing import Any

from jinja2 import Environment, FileSystemLoader, StrictUndefined, select_autoescape


class TemplateRenderer:
    """Render Jinja2 templates with workflow context."""

    def __init__(self, template_dirs: list[Path] | None = None):
        if template_dirs is None:
            default_templates = Path(__file__).parent.parent.parent.parent / "templates"
            template_dirs = [default_templates]

        self.template_dirs = template_dirs
        self.env = Environment(
            loader=FileSystemLoader([str(d) for d in template_dirs if d.exists()]),
            autoescape=select_autoescape(default=False),
            undefined=StrictUndefined,
        )
        self.env.globals = {}

    def render(self, template_name: str, context: dict[str, Any]) -> str:
        """Render a template file with context."""
        template = self.env.get_template(template_name)
        return template.render(**context)

    def render_string(self, template_str: str, context: dict[str, Any]) -> str:
        """Render a template string with context."""
        template = self.env.from_string(template_str)
        return template.render(**context)

    def has_variables(self, text: str) -> bool:
        """Check if text contains Jinja2 variables."""
        return "{{" in text or "{%" in text

    def add_template_dir(self, path: Path) -> None:
        """Add a template directory to the search path."""
        if path.exists() and path not in self.template_dirs:
            self.template_dirs.append(path)
            self.env.loader = FileSystemLoader(
                [str(d) for d in self.template_dirs if d.exists()]
            )


def build_template_context(
    workflow_name: str,
    started_at: str,
    completed_at: str | None,
    step_outputs: dict[str, Any],
    files_changed: list[str],
    branches: list[str],
    pull_requests: list[dict[str, Any]],
    inputs: dict[str, Any],
) -> dict[str, Any]:
    """Build the template context object for output templates."""
    return {
        "workflow": {
            "name": workflow_name,
            "started_at": started_at,
            "completed_at": completed_at,
        },
        "steps": step_outputs,
        "files_changed": files_changed,
        "branches": branches,
        "pull_requests": pull_requests,
        "inputs": inputs,
    }


def render_workflow_output(
    template_path: Path,
    output_path: Path,
    context: dict[str, Any],
    template_dirs: list[Path] | None = None,
) -> None:
    """Render a workflow output template to a file."""
    renderer = TemplateRenderer(template_dirs)

    if template_path.is_absolute():
        content = template_path.read_text(encoding="utf-8")
        rendered = renderer.render_string(content, context)
    else:
        rendered = renderer.render(str(template_path), context)

    output_path.parent.mkdir(parents=True, exist_ok=True)
    output_path.write_text(rendered, encoding="utf-8")

File: src/renderer.ts
/** Nunjucks template rendering for workflows. */

import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import nunjucks from "nunjucks";

const __dirname = path.dirname(fileURLToPath(import.meta.url));

/** Set of variable names already warned about (reset per render). */
const warnedVars = new Set<string>();

/**
 * Custom nunjucks loader that warns on undefined variables instead of failing.
 * Returns the original `{{ variable_name }}` syntax for undefined variables.
 */
function createLenientEnvironment(templateDirs: string[]): nunjucks.Environment {
	const loader = new nunjucks.FileSystemLoader(templateDirs, { noCache: true });
	const env = new nunjucks.Environment(loader, { autoescape: false });

	// Install a global that intercepts missing variables
	// Nunjucks doesn't have a direct WarnOnUndefined equivalent,
	// so we use a custom extension approach
	return env;
}

function createStrictEnvironment(templateDirs: string[]): nunjucks.Environment {
	const loader = new nunjucks.FileSystemLoader(templateDirs, { noCache: true });
	return new nunjucks.Environment(loader, {
		autoescape: false,
		throwOnUndefined: true,
	});
}

export class TemplateRenderer {
	templateDirs: string[];
	strictMode: boolean;
	private env: nunjucks.Environment;

	constructor(templateDirs?: string[], strictMode = false) {
		const dirs = templateDirs ?? [path.join(__dirname, "templates")];

		this.templateDirs = dirs;
		this.strictMode = strictMode;

		const existingDirs = dirs.filter((d) => existsSync(d));
		this.env = strictMode
			? createStrictEnvironment(existingDirs)
			: createLenientEnvironment(existingDirs);
	}

	render(templateName: string, context: Record<string, unknown>): string {
		if (!this.strictMode) {
			warnedVars.clear();
		}
		return this.env.render(templateName, context);
	}

	renderString(templateStr: string, context: Record<string, unknown>): string {
		if (!this.strictMode) {
			warnedVars.clear();
		}
		return this.env.renderString(templateStr, context);
	}

	hasVariables(text: string): boolean {
		return text.includes("{{") || text.includes("{%");
	}

	addTemplateDir(dirPath: string): void {
		if (existsSync(dirPath) && !this.templateDirs.includes(dirPath)) {
			this.templateDirs.push(dirPath);
			const existingDirs = this.templateDirs.filter((d) => existsSync(d));
			this.env = this.strictMode
				? createStrictEnvironment(existingDirs)
				: createLenientEnvironment(existingDirs);
		}
	}
}

// --- Helper functions ---

const ANALYSIS_TYPES = ["bug", "debt", "doc", "security", "style"];

export function extractAnalysisSteps(
	stepOutputs: Record<string, Record<string, unknown>>,
): Record<string, Record<string, unknown>> {
	const result: Record<string, Record<string, unknown>> = {};
	for (const [name, data] of Object.entries(stepOutputs)) {
		if (name.startsWith("analyze-")) {
			const analysisType = name.slice("analyze-".length);
			if (ANALYSIS_TYPES.includes(analysisType)) {
				result[name] = data;
			}
		}
	}
	return result;
}

export function extractFixSteps(
	stepOutputs: Record<string, Record<string, unknown>>,
): Record<string, Record<string, unknown>> {
	const result: Record<string, Record<string, unknown>> = {};
	for (const [name, data] of Object.entries(stepOutputs)) {
		if (name.startsWith("fix-") || name.startsWith("apply-")) {
			result[name] = data;
		}
	}
	return result;
}

export function buildTemplateContext(
	workflowName: string,
	workflowId: string,
	startedAt: string,
	completedAt: string | null,
	stepOutputs: Record<string, Record<string, unknown>>,
	filesChanged: string[],
	branches: string[],
	pullRequests: Record<string, unknown>[],
	inputs: Record<string, unknown>,
): Record<string, unknown> {
	const analysisSteps = extractAnalysisSteps(stepOutputs);
	const fixSteps = extractFixSteps(stepOutputs);

	return {
		workflow: {
			name: workflowName,
			id: workflowId,
			started_at: startedAt,
			completed_at: completedAt,
		},
		workflow_id: workflowId,
		steps: stepOutputs,
		analysis_steps: analysisSteps,
		fix_steps: fixSteps,
		files_changed: filesChanged,
		branches,
		pull_requests: pullRequests,
		inputs,
	};
}

export function renderWorkflowOutput(
	templatePath: string,
	outputPath: string,
	context: Record<string, unknown>,
	templateDirs?: string[],
	strictMode = false,
): void {
	const renderer = new TemplateRenderer(templateDirs, strictMode);

	let rendered: string;
	if (path.isAbsolute(templatePath)) {
		const content = readFileSync(templatePath, "utf-8");
		rendered = renderer.renderString(content, context);
	} else {
		rendered = renderer.render(templatePath, context);
	}

	const dir = path.dirname(outputPath);
	mkdirSync(dir, { recursive: true });
	writeFileSync(outputPath, rendered, "utf-8");
}
